storage/rocksdb/rdb_locking_iter.h

```diff
@@ -0,0 +1,190 @@
+//////////////////////////////////////////////////////////////////////////////
+
+//
+// LockingIterator is an iterator that locks the rows before returning, as well
+// as scanned gaps between the rows.
+//
+//  Example:
+//    lock_iter= trx->GetLockingIterator();
+//    lock_iter->Seek('abc');
+//    lock_iter->Valid()==true && lock_iter->key() == 'bcd';
+//
+//   After the above, the returned record 'bcd' is locked by transaction trx.
+//   Also, the range between ['abc'..'bcd'] is empty and locked by trx.
+//
+//    lock_iter->Next();
+//    lock_iter->Valid()==true && lock_iter->key() == 'efg'
+//
+//   Now, the range ['bcd'.. 'efg'] (bounds incluive) is also locked, and there are no
+//   records between 'bcd'  and 'efg'.
+//
+class LockingIterator : public rocksdb::Iterator {
+
+  rocksdb::Transaction *txn_;
+  rocksdb::ColumnFamilyHandle* cfh_;
+  bool m_is_rev_cf;
+  rocksdb::ReadOptions read_opts_;
+  rocksdb::Iterator *iter_;
+  rocksdb::Status status_;
+
+  // note: an iterator that has reached EOF has status()==OK && valid_==false
+  bool  valid_;
+
+  ulonglong *lock_count_;
+ public:
+  LockingIterator(rocksdb::Transaction *txn,
+                  rocksdb::ColumnFamilyHandle *cfh,
+                  bool is_rev_cf,
+                  const rocksdb::ReadOptions& opts,
+                  ulonglong *lock_count=nullptr
+                  ) :
+    txn_(txn), cfh_(cfh), m_is_rev_cf(is_rev_cf), read_opts_(opts), iter_(nullptr),
+    status_(rocksdb::Status::InvalidArgument()), valid_(false),
+    lock_count_(lock_count) {}
+
+  ~LockingIterator() {
+    delete iter_;
+  }
+
+  virtual bool Valid() const override { return valid_; }
+
+  // Note: MyRocks doesn't ever call these:
+  virtual void SeekToFirst() override;
+  virtual void SeekToLast() override;
+
+  virtual void Seek(const rocksdb::Slice& target) override;
+
+  // Position at the last key in the source that at or before target.
+  // The iterator is Valid() after this call iff the source contains
+  // an entry that comes at or before target.
+  virtual void SeekForPrev(const rocksdb::Slice& target) override;
+
+  virtual void Next() override;
+  virtual void Prev() override;
+
+  virtual rocksdb::Slice key() const override {
+    assert(Valid());
+    return iter_->key();
+  }
+
+  virtual rocksdb::Slice value() const override {
+    assert(Valid());
+    return iter_->value();
+  }
+
+  virtual rocksdb::Status status() const override {
+    return status_;
+  }
+
+ private:
+  template <bool forward> void Scan(const rocksdb::Slice& target,
+                                    bool call_next) {
+    if (!iter_->Valid()) {
+      status_ = iter_->status();
+      valid_ = false;
+      return;
+    }
+
+    while (1) {
+      /*
+        note: the underlying iterator checks iterator bounds, so we don't need
+        to check them here
+      */
+      DEBUG_SYNC(my_core::thd_get_current_thd(), "rocksdb.locking_iter_scan");
+      auto end_key = iter_->key();
```

Comment:
Would it might make sense to store the range locked so far. If this key falls within the range, we won't have to perform another GetRangeLock(). Thinking ahead, if we wanted to optimize this by scan ahead a few keys and locking the whole range first, we can reduce the number of GetRangeLock() calls.

Reply:
> If this key falls within the range, we won't have to perform another GetRangeLock().

Are you talking about the case where another transaction is inserting records into the range we're scanning? Unless I am missing something, this is the only scenario where the current code will make more GetRangeLock() calls than necessary.

I agree that if we scan ahead several values, we'll need to remember what range we've locked.

Reply:
Yes, I'm thinking of the case where our snapshot gives us a next_key and as we attempt to acquire a range lock up to next_key, a concurrent transaction inserts and commits multiple keys before our next_key. Once we lock up to next_key, we refresh the snapshot and discover the next few committed keys is already in the range lock up to next_key.

Reply:
Ok, done.